### PR TITLE
feature-erms-1901

### PR DESCRIPTION
**Pull request by croqueGrec09**
Full sync rework.

includes commits for:
- ERMS-1901: rework of sync procedure flow
- ERMS-1929/ERMS-1948/ERMS-1957/ERMS-2106: removal of legacy structures, cleanup of data, setting of gokbId foreign keys to unique and not null
- ERMS-1882: PendingChange behavior is configurable (tests needed to be done in QA phase)

### CODE REVIEW

**Comment by croqueGrec09** [comment]
Do NOT merge this pull request until EVERYONE approved it!

**Comment by croqueGrec09** [comment]
Further commits to expect, there are  bugfixes coming up!

File: app/migrations/changelog.groovy [hunk 1]
```diff
@@ -58,10 +58,10 @@ databaseChangeLog = {
 	include file: 'changelog-2020-02-06.groovy'
 	include file: 'changelog-2020-02-10.groovy'
 	include file: 'changelog-2020-02-18.groovy'
-	include file: 'changelog-2020-02-19.groovy'
 	include file: 'changelog-2020-03-02.groovy'
 	include file: 'changelog-2020-03-05.groovy'
 	include file: 'changelog-2020-03-09.groovy'
 	include file: 'changelog-2020-03-10.groovy'
 	include file: 'changelog-2020-03-13.groovy'
+	include file: 'changelog-2020-02-19.groovy' //because of other changes before, is for test
```
Comment (croqueGrec09) [inline app/migrations/changelog.groovy]: This is provisorical; David, we need to discuss whether this is legitime or if I should update the timestamp

Comment (croqueGrec09) [reply app/migrations/changelog.groovy]: outdated

File: app/deprecated/grails-app/controllers/com/k-int/kbplus/ProcessLoginController.groovy [hunk 2]
```diff
@@ -19,12 +17,12 @@ class ProcessLoginController {
   def index() {
     return
 
-    log.debug("ProcessLoginController::index() - session = ${request.session.id}");
+      com.k_int.kbplus.ProcessLoginController.log.debug("ProcessLoginController::index() - session = ${request.session.id}");
```
Comment (anjaalbinnrw) [inline app/deprecated/grails-app/controllers/com/k-int/kbplus/ProcessLoginController.groovy]: Warum importierst du die Klasse nicht?

File: app/deprecated/grails-app/controllers/com/k-int/kbplus/ProcessLoginController.groovy [hunk 3]
```diff
@@ -52,17 +49,17 @@ class ProcessLoginController {
...
-      log.debug("Inst Addr = ${map.authInstitutionAddress}");
+        com.k_int.kbplus.ProcessLoginController.log.debug("Auth inst = ${map.authInstitutionName}");
+        com.k_int.kbplus.ProcessLoginController.log.debug("UserId = ${map.eduPersonTargetedID}");
+        com.k_int.kbplus.ProcessLoginController.log.debug("email = ${map.mail}");
+        com.k_int.kbplus.ProcessLoginController.log.debug("Inst Addr = ${map.authInstitutionAddress}");
 
       if ( ( map.eduPersonTargetedID != null ) && ( map.eduPersonTargetedID.length() > 0 ) ) {
   
-        User user = User.findByUsername(map.eduPersonTargetedID)
+        def user = com.k_int.kbplus.auth.User.findByUsername(map.eduPersonTargetedID)
```
Comment (anjaalbinnrw) [inline app/deprecated/grails-app/controllers/com/k-int/kbplus/ProcessLoginController.groovy]: ?

Comment (croqueGrec09) [reply app/deprecated/grails-app/controllers/com/k-int/kbplus/ProcessLoginController.groovy]: Weil sie in Deprecated ist und Intellij hingegangen ist und alles importiert.

File: app/deprecated/grails-app/controllers/com/k-int/kbplus/ProcessLoginController.groovy (hunk 2 above)
Comment (croqueGrec09) [reply app/deprecated/grails-app/controllers/com/k-int/kbplus/ProcessLoginController.groovy]: s.u. - Klasse ist in Deprecated

File: app/grails-app/services/com/k_int/kbplus/ChangeNotificationService.groovy [hunk 4]
```diff
@@ -317,4 +326,65 @@ class ChangeNotificationService extends AbstractLockableService {
...
+                    settingValue = parentSP.pendingChangeConfig.find { PendingChangeConfiguration pcc -> pcc.settingKey == settingKey }.settingValue
+                }
+                if((settingValue == null && !subscriptionPackage.subscription.instanceOf) || settingValue == RDStore.PENDING_CHANGE_CONFIG_PROMPT) {
+                    //case four, then fallback or explicitly set as such
+                    PendingChange.construct([target:args.target,oid:args.oid,newValue:args.newValue,oldValue:args.oldValue,prop:args.prop,msgToken:msgToken,status:RDStore.PENDING_CHANGE_PENDING,owner:contextOrg])
+                }
+                else if(settingValue == RDStore.PENDING_CHANGE_CONFIG_ACCEPT) {
+                    //set up announcement and do accept! Pending because if some error occurs, the notification should still take place
+                    PendingChange pc = PendingChange.construct([target:args.target,oid:args.oid,newValue:args.newValue,oldValue:args.oldValue,prop:args.prop,msgToken:msgToken,status:RDStore.PENDING_CHANGE_PENDING,owner:contextOrg])
+                    pc.save()
```
Comment (klodav) [inline app/grails-app/services/com/k_int/kbplus/ChangeNotificationService.groovy]: PendingChange.construct() should save implicitly

Comment (croqueGrec09) [reply app/grails-app/services/com/k_int/kbplus/ChangeNotificationService.groovy]: Alright, will be changed ...

File: app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy [hunk 5]
```diff
@@ -117,13 +119,13 @@ class SubscriptionController extends AbstractDebugController {
     @DebugAnnotation(test = 'hasAffiliation("INST_USER")')
     @Secured(closure = { ctx.springSecurityService.getCurrentUser()?.hasAffiliation("INST_USER") })
     def index() {
-        def result = setResultGenericsAndCheckAccess(AccessService.CHECK_VIEW)
+        def result = setResultGenericsAndCheckAccess(accessService.CHECK_VIEW)
```
Comment (klodav) [inline app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy]: Still a class constant

Comment (croqueGrec09) [reply app/grails-app/controllers/com/k_int/kbplus/SubscriptionController.groovy]: Alright!

File: app/grails-app/controllers/com/k_int/kbplus/TitleController.groovy [hunk 6]
```diff
@@ -86,12 +86,16 @@ class TitleController extends AbstractDebugController {
...
     //def new_title = new TitleInstance(title:params.title, impId:java.util.UUID.randomUUID().toString()
-    def ti_status = RefdataValue.getByValueAndCategory('Current', RDConstants.TITLE_STATUS)
-    def new_title =  ((params.typ=='Ebook') ? new BookInstance(title:params.title, impId:java.util.UUID.randomUUID().toString(), status: ti_status, type: RefdataValue.getByValueAndCategory('EBook', RDConstants.TITLE_TYPE)) :
-              (params.typ=='Database' ? new DatabaseInstance(title:params.title, impId:java.util.UUID.randomUUID().toString(), status: ti_status, type: RefdataValue.getByValueAndCategory('Database', RDConstants.TITLE_TYPE)) : new JournalInstance(title:params.title, impId:java.util.UUID.randomUUID().toString(), status: ti_status, type: RefdataValue.getByValueAndCategory('Journal', RDConstants.TITLE_TYPE))))
+    def ti_status = RefdataValue.getByValueAndCategory('Current', RefdataCategory.TITLE_STATUS)
+    def new_title =  ((params.typ=='Ebook') ? new BookInstance(title:params.title, impId:java.util.UUID.randomUUID().toString(), status: ti_status, type: RefdataValue.getByValueAndCategory('EBook', RefdataCategory.TI_MEDIUM)) :
```
Comment (anjaalbinnrw) [inline app/grails-app/controllers/com/k_int/kbplus/TitleController.groovy]: Warum das?

Comment (croqueGrec09) [reply app/grails-app/controllers/com/k_int/kbplus/TitleController.groovy]: Ist legacy

File: app/grails-app/domain/com/k_int/kbplus/SubscriptionPackage.groovy [hunk 7]
```diff
@@ -22,9 +24,9 @@ class SubscriptionPackage {
     lastUpdated column: 'sp_last_updated'
   }
 
-
   static hasMany = [
-    oapls: OrgAccessPointLink
+    oapls: OrgAccessPointLink,
+    pendingChangeConfig: PendingChangeConfiguration
```
[the remark this answers is not in the document]
Comment (croqueGrec09) [reply app/grails-app/domain/com/k_int/kbplus/SubscriptionPackage.groovy]: Er wird entfernt. Entsprechende Anweisungen sind nachgereicht worden.

File: app/grails-app/views/myInstitution/dashboard.gsp [hunk 8]
```diff
@@ -7,7 +7,7 @@
         <meta name="layout" content="semanticUI"/>
         <title>${message(code:'laser')} : ${message(code:'menu.institutions.dash')}</title>
     </head>
-    <body>
+    <div>
```
Comment (klodav) [inline app/grails-app/views/myInstitution/dashboard.gsp]: Incorrect markup

Comment (croqueGrec09) [reply app/grails-app/views/myInstitution/dashboard.gsp]: Indeed ... will be fixed!

File: app/grails-app/views/issueEntitlement/show.gsp [hunk 9]
```diff
@@ -8,34 +8,34 @@
 </head>
 <body>
     <semui:breadcrumbs>
-        <g:if test="${issueEntitlementInstance?.subscription.subscriber}">
-            <semui:crumb controller="myInstitution" action="currentSubscriptions" params="${[shortcode:issueEntitlementInstance?.subscription.subscriber.shortcode]}" text="${issueEntitlementInstance?.subscription.subscriber.name} - ${message(code:'subscription.plural')}"/>
+        <g:if test="${issueEntitlementInstance.subscription.subscriber}">
+            <semui:crumb controller="myInstitution" action="currentSubscriptions" params="${[shortcode:issueEntitlementInstance?.subscription.subscriber.shortcode]}" text="${issueEntitlementInstance.subscription.subscriber.name} - ${message(code:'subscription.plural')}"/>
```
Comment (klodav) [inline app/grails-app/views/issueEntitlement/show.gsp]: to be or not to be -> ?

File: app/grails-app/views/issueEntitlement/show.gsp [hunk 10]
```diff
@@ -8,34 +8,34 @@
...
         </g:if>
-        <semui:crumb controller="subscription" action="index" id="${issueEntitlementInstance?.subscription.id}"  text="${issueEntitlementInstance?.subscription.name}" />
-        <semui:crumb class="active" id="${issueEntitlementInstance?.id}" text="${issueEntitlementInstance?.tipp.title.title}" />
+        <semui:crumb controller="subscription" action="index" id="${issueEntitlementInstance.subscription.id}"  text="${issueEntitlementInstance.subscription.name}" />
+        <semui:crumb class="active" id="${issueEntitlementInstance?.id}" text="${issueEntitlementInstance.tipp.title.title}" />
```
Comment (klodav) [inline app/grails-app/views/issueEntitlement/show.gsp]: to be or not to be -> ?

Comment (croqueGrec09) [reply app/grails-app/views/issueEntitlement/show.gsp]: forgotten ? ... to be!

File: app/grails-app/views/issueEntitlement/show.gsp (hunk 9 above)
Comment (croqueGrec09) [reply app/grails-app/views/issueEntitlement/show.gsp]: To be! The question mark shalt be forgotten ...

File: app/grails-app/views/dataManager/listDeletedTIPPS.gsp [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+  Date: 09.12.2019
+  Time: 15:25
+--%>
+
+<%@ page import="com.k_int.kbplus.IssueEntitlement; com.k_int.kbplus.TitleInstancePackagePlatform; grails.converters.JSON" contentType="text/html;charset=UTF-8" %>
+<laser:serviceInjection/>
+<html>
+    <head>
+        <meta name="layout" content="semanticUI"/>
+        <title>Expunge deleted TIPPs</title>
```
Comment (klodav) [inline app/grails-app/views/dataManager/listDeletedTIPPS.gsp]: <title>${message(code:'laser')} : Expunge deleted TIPPs</title>

File: app/grails-app/views/dataManager/listPlatformDuplicates.gsp [hunk 12]
```diff
@@ -0,0 +1,98 @@
+<%@ page import="com.k_int.kbplus.TitleInstance" %>
+<!doctype html>
+<html>
+  <head>
+    <meta name="layout" content="semanticUI"/>
+    <title>${message(code:'laser', default:'LAS:eR')} : Admin::Platform duplicates</title>
```
Comment (klodav) [inline app/grails-app/views/dataManager/listPlatformDuplicates.gsp]: ${message(code:'laser')}

File: app/grails-app/views/dataManager/listDeletedTIPPS.gsp (hunk 11 above)
Comment (croqueGrec09) [reply app/grails-app/views/dataManager/listDeletedTIPPS.gsp]: Corrected

File: app/grails-app/views/dataManager/listPlatformDuplicates.gsp (hunk 12 above)
Comment (croqueGrec09) [reply app/grails-app/views/dataManager/listPlatformDuplicates.gsp]: Corrected

File: app/grails-app/domain/com/k_int/kbplus/DatabaseInstance.groovy [hunk 13]
```diff
@@ -8,4 +10,15 @@ class DatabaseInstance extends TitleInstance{
 
     static constraints = {
     }
+
+    static DatabaseInstance construct(Map<String,Object> params) {
```
Comment (klodav) [inline app/grails-app/domain/com/k_int/kbplus/DatabaseInstance.groovy]: No implicit save?

File: app/grails-app/domain/com/k_int/kbplus/BookInstance.groovy [hunk 14]
```diff
@@ -47,6 +48,16 @@ class BookInstance extends TitleInstance {
 
     }
 
+    static BookInstance construct(Map<String,Object> params) {
```
Comment (klodav) [inline app/grails-app/domain/com/k_int/kbplus/BookInstance.groovy]: No implicit save?

File: app/grails-app/domain/com/k_int/kbplus/IssueEntitlement.groovy [hunk 15]
```diff
@@ -124,6 +126,17 @@ class IssueEntitlement extends AbstractBaseDomain implements Comparable {
     dateCreated (nullable: true, blank: false)
   }
 
+  static IssueEntitlement construct(Map<String,Object> configMap) throws EntitlementCreationException {
```
Comment (klodav) [inline app/grails-app/domain/com/k_int/kbplus/IssueEntitlement.groovy]: No implicit save?

File: app/grails-app/domain/com/k_int/kbplus/JournalInstance.groovy [hunk 16]
```diff
@@ -8,4 +10,15 @@ class JournalInstance extends TitleInstance {
 
     static constraints = {
     }
+
+    static JournalInstance construct(Map<String,Object> params) {
```
Comment (klodav) [inline app/grails-app/domain/com/k_int/kbplus/JournalInstance.groovy]: No implicit save?

File: app/grails-app/services/com/k_int/kbplus/GlobalSourceSyncService.groovy [hunk 17]
```diff
@@ -1,1728 +1,1012 @@
...
-                log.debug("Create new history event");
-                TitleHistoryEvent he = new TitleHistoryEvent(eventDate: query_params[0]).save()
-                fromset.each {
-                    new TitleHistoryEventParticipant(event: he, participant: it, participantRole: 'from').save()
-                }
-                toset.each {
-                    new TitleHistoryEventParticipant(event: he, participant: it, participantRole: 'to').save()
-                }
+            catch (Exception e) {
+                SystemEvent.createEvent('GSSS_OAI_ERROR',['jobId':source.id]).save()
```
Comment (klodav) [inline app/grails-app/services/com/k_int/kbplus/GlobalSourceSyncService.groovy]: Redundant save()

File: app/migrations/changelog-2020-02-19.groovy [hunk 18]
```diff
@@ -1,22 +0,0 @@
-databaseChangeLog = {
```
Comment (klodav) [inline app/migrations/changelog-2020-02-19.groovy]: file is required

File: app/migrations/changelog.groovy [hunk 19]
```diff
@@ -58,7 +58,6 @@ databaseChangeLog = {
 	include file: 'changelog-2020-02-06.groovy'
 	include file: 'changelog-2020-02-10.groovy'
 	include file: 'changelog-2020-02-18.groovy'
-	include file: 'changelog-2020-02-19.groovy'
```
Comment (klodav) [inline app/migrations/changelog.groovy]: required

**Pull request merged**